### ISSUE TITLE
Association Filters + Nested Drawers

### ISSUE THREAD

**Issue by jonmifsud** [symphonycms/symphony-2#2322]
I'm not exactly sure if this should be in the core but raising it here as I am noticing a few core related things. There's a [related ticket in the Associations Field](https://github.com/symphonists/association_field/issues/12) itself.

I would like to add filters for the Associations UI, I made some modifications to the extension itself and it can support filtering in the same way symphony publish pages do pretty easily. What I'd like some input on is two things: UI + core related modifications.

From what I can see so far the Associations Interface is made available using `appendAssociationInterfaceSelect` within the core `class.field.php` and checks if there is an Associations Interface and editor before showing the related options which is great, so in theory I think if the Associations Interface UI supported filters the UI has to be added in here theoretically.

Now coming to the UI related issue, I was hopping to basically drop in the same interface we have added for Publish Filtering in the core within the publish page. The screenshot below is a representation of how this would look like if dropped in "as is". Other than requiring an override of some css for the nested drawer + js to make the suggestions/autocomplete work it seems to work, except when you try to remove an object within the embedded drawer, as it closes the main drawer too, I think this is minor JS to avoid bubbling of the drawer events. What I think would be nice is to extract this filtering UI out of the publish page and making it re-usable. 

![screen shot 2015-01-16 at 09 37 48](https://cloud.githubusercontent.com/assets/859775/5773506/a7cf3562-9d63-11e4-9930-843fab66c193.png)

I'd appreciate some feedback as I will am hoping to wrap up this functionality in a couple of weeks.

**Comment by brendo** [comment]
> Now coming to the UI related issue, I was hopping to basically drop in the same interface we have added for Publish Filtering in the core within the publish page. The screenshot below is a representation of how this would look like if dropped in "as is". Other than requiring an override of some css for the nested drawer + js to make the suggestions/autocomplete work it seems to work, except when you try to remove an object within the embedded drawer, as it closes the main drawer too, I think this is minor JS to avoid bubbling of the drawer events. What I think would be nice is to extract this filtering UI out of the publish page and making it re-usable.

This would be a nice addition. I'm sure it doesn't work currently simply because it was not in scope of the Publish Filtering changes. It was never envisaged to reuse the interface elsewhere. Nice to know it's almost there though :)

**Comment by nilshoerrmann** [comment]
Shouldn't be that hard to fix, I guess.

**Comment by jonmifsud** [comment]
No I don't think so, I will probably open a PR to track progress in that regard. Just didn't want to proceed in case I was taking the wrong approach.

**Comment by jonmifsud** [comment]
@brendo / @nilshoerrmann do you think it's safe to append `event.stopPropagation();` to each callback for the duplicator and collapsable interfaces + backend.views.js

I've fixed most of the duplicator collapsable issues by stopping event propagation but not sure if I should be testing each event or if it's safe to stop propagation for each item. With propagation nested duplicators create all these weird effects. By default `.trigger` functions bubble up the dom I'm assuming this was never an issue they were not nested.

**Comment by nilshoerrmann** [comment]
> do you think it's safe to append event.stopPropagation(); to each callback for the duplicator and collapsable interfaces + backend.views.js

No, I don't think so.
Might work if applied to nested Duplicators only (Duplicator in Duplicator).

**Comment by jonmifsud** [comment]
Hmm.. is there a set of tests I could run somewhere to make sure I don't break anything then? (Can be manual testing)

So far within the section editor I have no issues with stopping propagation. Propagation only stops per event function each duplicator callback is tied to `.instance` so I think it's unlikely that you want to re-run those callbacks in another `.instance` higher up the dom..

**Comment by nilshoerrmann** [comment]
Well, testing the UI … that's really not a strenght of Symphony. A lot of room for improvement here.

Check out other areas of the system and extensions:

- the new publish interface
- the JIT preferences
- the meta keys field
- the date and time field
- the data source editor

At least, stopping propagation is nothing I'd add frivolously.

**Comment by nitriques** [comment]
> At least, stopping propagation is nothing I'd add frivolously.

Me neither. Some extensions (like DateTime) relies on the propagation in order to achieve their goals.

It's not the best architecture, but it the one we have now and it works OK.

**Comment by jonmifsud** [comment]
Makes sense - I'm going through trying to find which propagations were causing an issue for me. From there we can figure out the next steps.06

**Comment by nitriques** [comment]
Great.

**Comment by jonmifsud** [comment]
These two seem the main culprits:

1. [Collapse Collapsible](https://github.com/symphonycms/symphony-2/blob/master/symphony/assets/js/src/symphony.collapsible.js#L97) collapses the main drawer when trying to remove/destroy a sub drawer item.

2. [Expand Collapsable](https://github.com/symphonycms/symphony-2/blob/master/symphony/assets/js/src/backend.views.js#L421) when trying to create a new item it animates / scrolls upwards to some strange place.

I think with those two the ui for sub-nav seems to work fine. I'll keep digging into this next week, but looks promising.

**Comment by nitriques** [comment]
>  I'll keep digging into this next week, but looks promising.

Great. 

> trying to create a new item it animates / scrolls upwards to some strange place.

This is related to focus on select tags. 

See https://code.google.com/p/chromium/issues/detail?id=423155
https://github.com/symphonycms/selectbox_link_field/issues/79
https://github.com/symphonycms/symphony-2/pull/2231

**Comment by jonmifsud** [comment]
Will check if related but doing a stop propagation on both of those events
gets it to work

**Comment by nitriques** [comment]
By curiousity, do you have the DateTime extension installed ? It's the only extension that could break I can think of...

**Comment by jonmifsud** [comment]
> This is related to focus on select tags.

Correct had a look at it basically it focuses on the first input field within the duplicator instance. Stopping propagation puts the focus on the search box as shown below. If I let the propagation bubble it focuses on the `Label: Section` at the top of the first duplicator instance.

![screen shot 2015-01-24 at 07 17 27](https://cloud.githubusercontent.com/assets/859775/5886330/4d9c7aa0-a399-11e4-9ce5-0020a53b702c.png)

DateTime seems to be working fine other than the messed up css on integration, due to the calendar stuff.

**Comment by jonmifsud** [comment]
@nitriques & @brendo does it make sense to try move out `createFilteringDuplicator` from content.publish.php into Widgets? or would it best be it's own class?

I think if we're making it includable in other elements other than the publish page as I intend to do, it doesn't make much sense to reside there. Also in the above use-case I don't want it to reside in a separate form, but rather be part of the page flow.

**Comment by nitriques** [comment]
>  does it make sense to try move out createFilteringDuplicator from content.publish.php into Widgets?

I really think it does... @brendo where would you put it ?

> DateTime seems to be working fine other than the messed up css on integration

Do you have a PR ready that I could test ??? Or just a branch ?

**Comment by jonmifsud** [comment]
Just Local. I will move this to Widgets and push the branch. Today was AFK
most of the day so a bit behind but I'll try get it done later on or
tomorrow.

**Comment by nitriques** [comment]
Awesome, thanks Jonathan!

**Comment by jonmifsud** [comment]
@nitriques [there you go](https://github.com/jonmifsud/symphony-2/tree/association-filters). Should I open a PR mark it pending? or finish all functionality first.

Next step I guess is where to store the filtering data... I have modified the [Association UI Selector in a branch filter](https://github.com/jonmifsud/association_ui_selector/tree/filters) which handles filtering using data attributes. So one can modify via JS, say filtering options change on parameters (what I eventually have in mind)

Any suggestions? I'm thinking possibly a new table linked to `tbl_sections_association` but don't want to overcomplicate things.

**Comment by nitriques** [comment]
Looks good. Hope to find the time to test ASAP.

Will you be able to send two separate PR ?

**Comment by jonmifsud** [comment]
Yes sure no problem. I guess you mean one to add the filters within the context and another PR for the rest? Or one for UI + other for storage functionality?

**Comment by nitriques** [comment]
No, sorry, I thought it was two separate things, but it's not. Sorry! One PR will be ok.

**Comment by brendo** [comment]
> @nitriques & @brendo does it make sense to try move out createFilteringDuplicator from content.publish.php into Widgets? or would it best be it's own class?

Sorry, I've been on holidays and missed this. It should be it's own class. It's fairly specific and has dependencies.

**Comment by nitriques** [comment]
Jonathan, did you had the time to work on this ?

**Comment by jonmifsud** [comment]
Partially didn't have time to wrap it up. Had to put on back burner as
project is a bit behind. I'll give it another look tomorrow.

I think UI works just need to save the post data into a database table of
some sort.

**Comment by nitriques** [comment]
Thanks.

**Comment by jonmifsud** [comment]
@nitriques did an update on my branch, happy to start a PR marked pending some further alterations.

1. I need to update the `sym_sections_association` table to add a new field - not sure where I should add it within the migrations - if it will be merged for 3.0 maybe it can go in a new file there

    ALTER TABLE `sym_sections_association`
    ADD `filters` text COLLATE utf8_unicode_ci;

2. Currently only 1 set of filtering is supported. I'd like to iterate through this some point maybe later this week to support conditional filtering. Eg: if one has multiple sections linked via a single association, there can be different filters per associated sections + filters could also change depending on the values of the current object. I feel like the UI could get too cluttered if not done properly.

**Comment by nitriques** [comment]
Great. And yes this would have to wait for 3.0.0. Thanks!
